### PR TITLE
Update sqlparser requirement from 0.38.0 to 0.39.0

### PR DESCRIPTION
**Pull request by jackwener**
## Which issue does this PR close?

Closes #7980.

## Rationale for this change

Update sqlparser requirement from 0.38.0 to 0.39.0

### CODE REVIEW

**Comment by jackwener** [comment]
sqlparser-0.39 parse `struct` as `EXPR`, but sqlparser-0.39 parse `struct` as `Function`.
So datafusion parse `struct` need to refactor.

File: datafusion/sql/src/expr/mod.rs [hunk 1]
```diff
@@ -477,10 +482,36 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
 
+            SQLExpr::Struct { values, fields } => {
+                self.parse_struct(values, fields, schema, planner_context)
+            }
+
             _ => not_impl_err!("Unsupported ast node in sqltorel: {sql:?}"),
         }
     }
 
+    fn parse_struct(
```
Comment (jackwener) [inline datafusion/sql/src/expr/mod.rs:493]: We also need to remove original code that parse `struct` as function.

**Comment by alamb** [comment]
I pushed a fix for clippy -- but there is still something wrong with FILTER aggregates.

Likely related to https://github.com/sqlparser-rs/sqlparser-rs/pull/1007

**Comment by jackwener** [comment]
Now there's a problem that parse hive-dialect `filter (XX)`.

File: datafusion/sql/src/expr/function.rs [hunk 2]
```diff
@@ -146,9 +146,14 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
                     self.function_args_to_expr(function.args, schema, planner_context)?;
+                let filter: Option<Box<Expr>> = function
+                    .filter
+                    .map(|e| self.sql_expr_to_logical_expr(*e, schema, planner_context))
+                    .transpose()?
+                    .map(Box::new);
 
                 return Ok(Expr::AggregateFunction(expr::AggregateFunction::new(
-                    fun, args, distinct, None, order_by,
+                    fun, args, distinct, filter, order_by,
```
Comment (alamb) [inline datafusion/sql/src/expr/function.rs:156]: 👍

**Review by alamb — APPROVED** [review]
Thank you @jackwener 🙏

**Pull request merged**